acvm/src/pwg.rs

```diff
@@ -374,4 +426,375 @@ mod test {
+        let PartialWitnessGeneratorStatus::RequiresOracleData { unsolved_opcodes, mut unresolved_brilligs, .. } = solver_status else {
+            panic!("Should require oracle data")
+        };
+
+        assert_eq!(unsolved_opcodes.len(), 0, "brillig should have been removed");
+        assert_eq!(unresolved_brilligs.len(), 1, "should have a brillig oracle request");
+
+        let UnresolvedBrillig { foreign_call_wait_info, mut brillig } =
+            unresolved_brilligs.remove(0);
+        assert_eq!(foreign_call_wait_info.inputs.len(), 1, "Should be waiting for a single input");
+
+        let x_plus_y_inverse = foreign_call_wait_info.inputs[0].inner.inverse();
+        // Alter Brillig oracle opcode
+        brillig
+            .foreign_call_results
+            .push(ForeignCallResult { values: vec![Value::from(x_plus_y_inverse)] });
+
+        let mut next_opcodes_for_solving = vec![Opcode::Brillig(brillig)];
+        next_opcodes_for_solving.extend_from_slice(&unsolved_opcodes[..]);
+        // After filling data request, continue solving
+        let solver_status =
+            pwg::solve(&backend, &mut witness_assignments, &mut blocks, next_opcodes_for_solving)
+                .expect("should stall on oracle");
+        let PartialWitnessGeneratorStatus::RequiresOracleData { unsolved_opcodes, mut unresolved_brilligs, .. } = solver_status else {
+            panic!("Should require oracle data")
+        };
+
+        assert!(unsolved_opcodes.is_empty(), "should be fully solved");
+        assert_eq!(unresolved_brilligs.len(), 1, "should have no unresolved oracles");
+
+        let UnresolvedBrillig { foreign_call_wait_info, mut brillig } =
+            unresolved_brilligs.remove(0);
+        assert_eq!(foreign_call_wait_info.inputs.len(), 1, "Should be waiting for a single input");
+
+        let i_plus_j_inverse = foreign_call_wait_info.inputs[0].inner.inverse();
+        assert_ne!(x_plus_y_inverse, i_plus_j_inverse);
+        // Alter Brillig oracle opcode
+        brillig
+            .foreign_call_results
+            .push(ForeignCallResult { values: vec![Value::from(i_plus_j_inverse)] });
+        let mut next_opcodes_for_solving = vec![Opcode::Brillig(brillig)];
+        next_opcodes_for_solving.extend_from_slice(&unsolved_opcodes[..]);
+
+        // After filling data request, continue solving
+        let solver_status =
+            pwg::solve(&backend, &mut witness_assignments, &mut blocks, next_opcodes_for_solving)
+                .expect("should not stall on oracle");
+        assert_eq!(solver_status, PartialWitnessGeneratorStatus::Solved, "should be fully solved");
+    }
+
+    #[test]
+    fn brillig_oracle_predicate() {
+        // Opcodes below describe the following:
+        // fn main(x : Field, y : pub Field, cond: bool) {
+        //     let z = x + y;
+        //     let z_inverse = 1/z
+        //     if cond {
+        //         constrain z_inverse == Oracle("inverse", x + y);
```

Comment:
```suggestion
        //         assert( z_inverse == Oracle("inverse", x + y) );
```

Reply:
Even if constrain was available in Noir, I'd opt to remove these since it can be confusing using `constrain` in the context of unconstrained functions